### PR TITLE
Fixes 4526: added subscription check in templates

### PR DESCRIPTION
**Pull request by dominikvagner**
## Summary
This PR introduces a subscription check for disabling template add/edit actions (and routes) for users that don't have a valid RHEL subscription.

## Testing steps
Verify that when signed in with a user that has a valid subscription everything works fine. \
And when signed in with a user that doesn't (the future API implementation determines that) these things are true:
- _Add Content Template_ is disabled on the template page.
- The kebab menu is disabled for each item in the template table.
- The actions select in the _Template Detail_ view is disabled.
- You can't access the add/edit template modal by visiting the URL for it.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-4526

File: src/Pages/Templates/TemplateDetails/components/TemplateActionDropdown.tsx [hunk 1]
```diff
@@ -49,15 +49,23 @@ export default function TemplateActionDropdown() {
     }
   };
 
+  const hasRHELSubscription = subscriptions?.red_hat_enterprise_linux || false;
+  const isMissingRequirements = !rbac?.templateWrite || !hasRHELSubscription;
+  const missingRequirements: string = !rbac?.templateWrite
+    ? 'permission'
+    : !hasRHELSubscription
+      ? 'subscription (RHEL)'
+      : 'permission';
```
Comment (Andrewgdewar) [inline src/Pages/Templates/TemplateDetails/components/TemplateActionDropdown.tsx:58]: Would the following not simplify this:
`rbac?.templateWrite && !hasRHELSubscription ? 'subscription (RHEL)' : 'permission'`

File: src/Pages/Templates/TemplatesTable/components/TemplateFilters.tsx [hunk 2]
```diff
@@ -65,6 +65,14 @@ const Filters = ({ isLoading, setFilterData, filterData }: Props) => {
   const { distribution_arches = [], distribution_versions = [] } =
     queryClient.getQueryData<RepositoryParamsResponse>(REPOSITORY_PARAMS_KEY) || {};
 
+  const hasRHELSubscription = subscriptions?.red_hat_enterprise_linux || false;
+  const isMissingRequirements = !rbac?.templateWrite || !hasRHELSubscription;
+  const missingRequirements: string = !rbac?.templateWrite
```
Comment (Andrewgdewar) [inline src/Pages/Templates/TemplatesTable/components/TemplateFilters.tsx:70]: [ditto](https://github.com/content-services/content-sources-frontend/pull/328/files#r1747465615)

File: src/services/Subscriptions/SubscriptionQueries.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { useState } from 'react';
+import { Subscriptions, getSubscriptions } from './SubscriptionApi';
+import useErrorNotification from 'Hooks/useErrorNotification';
+
+export const useFetchSubscriptionsQuery = () => {
+  const [isLoading, setIsLoading] = useState(false);
+  const errorNotifier = useErrorNotification();
+
+  const fetchSubscriptions = async (): Promise<Subscriptions | null> => {
```
Comment (Andrewgdewar) [inline src/services/Subscriptions/SubscriptionQueries.ts:9]: Use react query here.
Although it isn't really needed for caching, it is used at the time level for monitoring of API calls for QE test reference. 
You can see how that works [here](https://github.com/content-services/content-sources-frontend/blob/main/src/App.tsx#L64).

File: src/middleware/AppContext.tsx [hunk 4]
```diff
@@ -32,6 +36,8 @@ export const ContextProvider = ({ children }: { children: ReactNode }) => {
   const chrome = useChrome();
   const [contentOrigin, setContentOrigin] = useState<ContentOrigin>(ContentOrigin.CUSTOM);
   const { fetchFeatures, isLoading: isFetchingFeatures } = useFetchFeaturesQuery();
+  const [subscriptions, setSubscriptions] = useState<Subscriptions | null>(null);
+  const { fetchSubscriptions, isLoading: isFetchingSubscriptions } = useFetchSubscriptionsQuery();
```
Comment (Andrewgdewar) [inline src/middleware/AppContext.tsx:40]: By using react query we can just use their in built hooks to take care of the state, so we won't need to bloat the state/useEffect here at all.

File: src/Pages/Templates/TemplatesTable/TemplatesTable.tsx [hunk 5]
```diff
@@ -154,7 +154,8 @@ const TemplatesTable = () => {
     versionDisplay,
   } = useArchVersion();
 
-  const actionTakingPlace = isLoading || isFetching || repositoryParamsLoading || isDeleting;
+  const actionTakingPlace =
+    isLoading || isFetching || repositoryParamsLoading || isDeleting || isFetchingSubscriptions;
```
Comment (Andrewgdewar) [inline src/Pages/Templates/TemplatesTable/TemplatesTable.tsx:158]: We wouldn't add isFetchingSubscriptions anywhere other than the top level app.tsx

File: src/Pages/Templates/TemplatesTable/TemplatesTable.tsx [hunk 6]
```diff
@@ -170,6 +171,14 @@ const TemplatesTable = () => {
 
   const countIsZero = count === 0;
 
+  const hasRHELSubscription = subscriptions?.red_hat_enterprise_linux || false;
```
Comment (Andrewgdewar) [inline src/Pages/Templates/TemplatesTable/TemplatesTable.tsx:174]: If you want a false here, just "double bang"
`!!subscriptions?.red_hat_enterprise_linux`

File: src/middleware/AppContext.tsx [hunk 7]
```diff
@@ -66,6 +77,8 @@ export const ContextProvider = ({ children }: { children: ReactNode }) => {
         rbac: rbac,
         features: features,
         isFetchingFeatures: isFetchingFeatures,
+        subscriptions: subscriptions,
+        isFetchingSubscriptions: isFetchingSubscriptions,
```
Comment (Andrewgdewar) [inline src/middleware/AppContext.tsx:81]: We can likely combine isFetchingFeatures and isFetchingSubscriptions into something like:
"isFetchingPermissions" for use downstream.
No need for two variables.

Comment (dominikvagner) [reply src/middleware/AppContext.tsx]: good point 👍🏼 resolved [here](https://github.com/content-services/content-sources-frontend/pull/328/commits/92ada247ce4e2cfe19fb4e6adac241f9ae08e54d) 🫡

File: src/Pages/Templates/TemplatesTable/TemplatesTable.tsx (hunk 6 above)
Comment (dominikvagner) [reply src/Pages/Templates/TemplatesTable/TemplatesTable.tsx]: didn't know bang/"double bang" can/should be used like that 😮 resolved [here](https://github.com/content-services/content-sources-frontend/pull/328/commits/92ada247ce4e2cfe19fb4e6adac241f9ae08e54d) 🫡

File: src/Pages/Templates/TemplatesTable/TemplatesTable.tsx (hunk 5 above)
Comment (dominikvagner) [reply src/Pages/Templates/TemplatesTable/TemplatesTable.tsx]: ok, resolved [here](https://github.com/content-services/content-sources-frontend/pull/328/commits/92ada247ce4e2cfe19fb4e6adac241f9ae08e54d) 🫡

File: src/services/Subscriptions/SubscriptionQueries.ts (hunk 3 above)
Comment (dominikvagner) [reply src/services/Subscriptions/SubscriptionQueries.ts]: sure, good point 👍🏼 I just made it similar to `useFetchFeaturesQuery` which I guess is outdated 😅 resolved [here](https://github.com/content-services/content-sources-frontend/pull/328/commits/92ada247ce4e2cfe19fb4e6adac241f9ae08e54d) 🫡

File: src/Pages/Templates/TemplateDetails/components/TemplateActionDropdown.tsx (hunk 1 above)
Comment (dominikvagner) [reply src/Pages/Templates/TemplateDetails/components/TemplateActionDropdown.tsx]: resolved [here](https://github.com/content-services/content-sources-frontend/pull/328/commits/92ada247ce4e2cfe19fb4e6adac241f9ae08e54d) 🫡

File: src/services/Subscriptions/SubscriptionQueries.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { Subscriptions, getSubscriptions } from './SubscriptionApi';
+import useErrorNotification from 'Hooks/useErrorNotification';
+import { useQuery } from 'react-query';
+
+const SUBSCRIPTION_CHECK_KEY = 'SUBSCRIPTION_CHECK_KEY';
+
+export const useFetchSubscriptionsQuery = () => {
+  const errorNotifier = useErrorNotification();
+  const { data: fetchSubscriptions, isLoading } = useQuery<Subscriptions>(
```
Comment (Andrewgdewar) [inline src/services/Subscriptions/SubscriptionQueries.ts:9]: Simply returning the useQuery directly should likely be done here.
In that way we can easily access any of the other variables (like isFetching, isError, error) should we wish to.

**Review by Andrewgdewar — COMMENTED** [review]
A few small #nit changes, but looks good. 
I'll do a functional check when this is marked as ready for review.

File: src/services/Subscriptions/SubscriptionQueries.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    [SUBSCRIPTION_CHECK_KEY],
+    () => getSubscriptions(),
+    {
+      keepPreviousData: true,
+      optimisticResults: true,
+      staleTime: 60000,
```
[the remark this answers is not in the document]
Comment (dominikvagner) [reply src/services/Subscriptions/SubscriptionQueries.ts]: okey dokey, removed that 🫡

File: src/services/Subscriptions/SubscriptionQueries.ts (hunk 8 above)
Comment (dominikvagner) [reply src/services/Subscriptions/SubscriptionQueries.ts]: good point, it should be done like that 👍🏼 fixed 😄

**Comment by Andrewgdewar** [comment]
Okay functional review looks good, just a question here. 
(More for @jlsherrill) 
Should we be using this functionality for more than just the templating-write permissions. 
Would it make sense to stop a user from associating a template to a system as well? 
Screenshot 2024-09-11 at 2 55 49 PM

**Comment by jlsherrill** [comment]
Yes, associating or unassociating a system should require template write (for now)

**Comment by Andrewgdewar** [comment]
Adding another requirement here! 😅 
All template edit related actions  should be disabled if the user does not have a subscription.
But the delete action should still be permitted if the user has template-edit permissions.

This would mean that we should enable both of these drop downs: 
Screenshot 2024-09-19 at 6 35 58 PM
Screenshot 2024-09-19 at 6 36 16 PM

But disable the corresponding actions within them if the user does not have a subscription.

**Review by Andrewgdewar — APPROVED** [review]
Ack!

**Comment by swadeley** [comment]
Hi

Thanks for the fix @dominikvagner 

No longer breaks ephemeral, merging and  will test again in stage.

**Pull request merged**